packages/integration-karma/test/rendering/iteration/index.spec.js

```diff
@@ -0,0 +1,59 @@
+import { extractDataIds } from 'test-utils';
+import { createElement } from 'lwc';
+import ForEach from 'x/forEach';
+import Iterator from 'x/iterator';
+import InlineForEach from 'x/inlineForEach';
+import InlineIterator from 'x/inlineIterator';
+
+describe('iteration rendering', () => {
+    function validateRenderedChildren(elm, iterationType) {
+        const isIterator = iterationType === 'iterator';
+        const expectedChildren = elm.items;
+        const { children } = elm.shadowRoot;
+        expect(children.length).toEqual(expectedChildren.length);
+
+        for (let i = 0; i < children.length; i++) {
+            const child = children[i];
+            const expected = expectedChildren[i];
+            const dataIds = extractDataIds(child);
+            expect(dataIds.first.textContent).toEqual(isIterator ? (i === 0).toString() : '');
+            expect(dataIds.last.textContent).toEqual(
+                isIterator ? (i === children.length - 1).toString() : ''
+            );
+            expect(dataIds.index.textContent).toEqual(i.toString());
+            expect(dataIds.value.textContent).toEqual(expected.toString());
+        }
+    }
+
+    const components = {
+        'x-inline-for-each': InlineForEach,
+        'x-inline-iterator': InlineIterator,
+        'x-template-for-each': ForEach,
+        'x-template-iterator': Iterator,
+    };
+
+    ['inline'].forEach((type) => {
+        ['iterator'].forEach((iterationType) => {
+            const tag = `x-${type}-${iterationType}`;
+            const Ctor = components[tag];
+
+            it(`${type} ${iterationType}`, () => {
+                const elm = createElement(tag, { is: Ctor });
+                elm.items = [1, 2, 3, 4];
+                document.body.appendChild(elm);
+                validateRenderedChildren(elm, iterationType);
+                const [c1, c2, c3, c4] = [...elm.shadowRoot.children];
+
+                elm.items = [3, 1, 2, 4];
+                return Promise.resolve().then(() => {
+                    validateRenderedChildren(elm, iterationType);
+                    const [c3b, c1b, c2b, c4b] = [...elm.shadowRoot.children];
+                    expect(c1).toBe(c1b);
+                    expect(c2).toBe(c2b);
+                    expect(c3).toBe(c3b);
+                    expect(c4).toBe(c4b);
+                });
+            });
+        });
+    });
+});
```

Comment:
This test passes in `master`. On 236 (v2.5.13) and 234 (v2.2.12), it throws an error:

    Error: Invariant Violation: Compiler should produce html functions that always return an array.

The reason for the error is that the `tmpl` function is returning `api_custom_element` rather than `api_iterator`. So it looks like previously we didn't actually support inline iterators, only inline `for:each`.